File: code/py/src/service_client/core/service_client.py
"""Transport-agnostic service client interface, config, request, and response types."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, Protocol
from urllib.parse import urlparse, urlunparse


class ServiceProtocol(StrEnum):
    """Transport protocol used by a ServiceRequest.

    Currently only HTTP is implemented. Future versions may add
    additional protocols (gRPC, WebSocket, message queues).
    """

    HTTP = "http"


@dataclass(frozen=True, slots=True)
class ServiceClientConfig:
    """Configuration for a ServiceClient: base URL, default headers, and timeout.

    The base_url path is normalized to end with '/' for correct endpoint resolution.
    Timeout is in seconds (default: 30).
    """

    base_url: str
    default_headers: dict[str, str] = field(default_factory=dict)
    timeout: float = 30.0

    def __post_init__(self) -> None:
        # frozen=True requires object.__setattr__ for post-init normalization
        normalized = self._normalize_base_url(self.base_url)
        object.__setattr__(self, "base_url", normalized)

    @staticmethod
    def _normalize_base_url(url: str) -> str:
        """Ensure the path ends with '/' for correct endpoint resolution."""
        parsed = urlparse(url)
        path = parsed.path
        if path == "" or path.endswith("/"):
            return url
        normalized = parsed._replace(path=f"{path}/")
        return urlunparse(normalized)


@dataclass(frozen=True, slots=True)
class ServiceRequest:
    """Describes a request to send through a ServiceClient."""

    protocol: ServiceProtocol
    method: str
    endpoint: str
    headers: dict[str, str] | None = None
    body: object | None = None
    error_message: str = "Error in service request"

    @staticmethod
    def http(
        *,
        method: str,
        endpoint: str,
        headers: dict[str, str] | None = None,
        body: object | None = None,
        error_message: str = "Error in HTTP request",
    ) -> ServiceRequest:
        """Factory for HTTP requests."""
        return ServiceRequest(
            protocol=ServiceProtocol.HTTP,
            method=method,
            endpoint=endpoint,
            headers=headers,
            body=body,
            error_message=error_message,
        )


@dataclass(frozen=True, slots=True)
class ServiceResponse:
    """Response returned by a ServiceClient after a successful send."""

    status_code: int
    headers: dict[str, str]
    raw_body: str
    data: Any = None


class ServiceClient(Protocol):
    """Transport-agnostic interface for sending service requests."""

    async def send(self, request: ServiceRequest) -> ServiceResponse: ...

    async def close(self) -> None: ...

File: code/ts/src/core/service_client.ts
/// Transport protocol used by a ServiceRequest.
///
/// Currently only 'http' is implemented. Future versions may add
/// additional protocols (gRPC, WebSocket, message queues).
export type ServiceProtocol = 'http';

export class ServiceClientConfig {
  readonly baseUrl: URL;
  readonly defaultHeaders: Record<string, string>;
  readonly timeout: number;

  constructor(opts: {
    baseUrl: URL | string;
    defaultHeaders?: Record<string, string>;
    /** Timeout in milliseconds. Default: 30000 (30s). */
    timeout?: number;
  }) {
    this.baseUrl = ServiceClientConfig._normalizeBaseUrl(
      typeof opts.baseUrl === 'string' ? new URL(opts.baseUrl) : opts.baseUrl,
    );
    this.defaultHeaders = opts.defaultHeaders ?? {};
    this.timeout = opts.timeout ?? 30_000;
  }

  /// Ensure the path ends with `/` for correct endpoint resolution.
  private static _normalizeBaseUrl(url: URL): URL {
    const path = url.pathname;
    if (path === '' || path.endsWith('/')) {
      return url;
    }
    const normalized = new URL(url.toString());
    normalized.pathname = `${path}/`;
    return normalized;
  }
}

export class ServiceRequest {
  readonly protocol: ServiceProtocol;
  readonly method: string;
  readonly endpoint: string;
  readonly headers?: Record<string, string>;
  readonly body?: unknown;
  readonly errorMessage: string;

  constructor(opts: {
    protocol: ServiceProtocol;
    method: string;
    endpoint: string;
    headers?: Record<string, string>;
    body?: unknown;
    errorMessage?: string;
  }) {
    this.protocol = opts.protocol;
    this.method = opts.method;
    this.endpoint = opts.endpoint;
    this.headers = opts.headers;
    this.body = opts.body;
    this.errorMessage = opts.errorMessage ?? 'Error in service request';
  }

  static http(opts: {
    method: string;
    endpoint: string;
    headers?: Record<string, string>;
    body?: unknown;
    errorMessage?: string;
  }): ServiceRequest {
    return new ServiceRequest({
      protocol: 'http',
      method: opts.method,
      endpoint: opts.endpoint,
      headers: opts.headers,
      body: opts.body,
      errorMessage: opts.errorMessage ?? 'Error in HTTP request',
    });
  }
}

export interface ServiceResponse {
  readonly statusCode: number;
  readonly headers: Record<string, string>;
  readonly rawBody: string;
  readonly data?: unknown;
}

export interface ServiceClient {
  send(request: ServiceRequest): Promise<ServiceResponse>;
  close(): Promise<void>;
}
